### ISSUE TITLE
How can I increase batch_size without memory error ?

### ISSUE THREAD

**Issue by unrealwill** [fchollet/keras#3917]
I have a gradient which is noisy, so it needs to be accumulated across many examples, without moving. 
It converges better when I increase batch_size (~1000) but it consumes more memory. 

When using theano without keras, I used to accumulate the gradient (in an extra accumulated gradient variable) from multiple passes over the examples. 
Memory usage could be reduced (a lot) by performing multiple passes. It could also allow some multi-gpu parallelism.

Is there a mechanism in Keras for doing so ?

(The only problem I see is with ops like Batch Normalization which can't be calculated easily with reduced memory).

Thank you

**Comment by robertomest** [comment]
You could compute the gradients multiple times and average them, like you were doing, but you would have to write your own training code. The keras backend has a gradient [function that computes gradients](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L730) much like Theano, so you could take a look at that if you are interested.
Could I ask what model you are training? Because batches of size 1000 are quite big. The noise in the gradient can many times be beneficial and having to do so many passes to make one step will probably make your training a lot slower.

**Comment by unrealwill** [comment]
Thank you, I will give a look this week-end. It would be great if I could reuse the keras training code. 
I'm currently training some RNN with a CTC loss function while trying to keep the number of parameters (and number of layers) quite low to be able to use it in an embedded device.

But I think the feature I request would be a useful addition because quite often there are some differences between GPUs we use in the lab, and GPUs in the cloud in term of memory. It would be cool to be able to lift this memory restriction when we need to be able to trade time for memory. (Quite often some people don't have 12G GPUs so may not be able to reproduce some results.

This is just averaging the gradient, but I'm not expert in keras enough to know where it would be better to implement this. I see 2 options, either wrapping the function, or wrapping the optimizers. But a Keras expert could probably do this a lot more efficiently than I would.

**Comment by robertomest** [comment]
Have you tried simply using a smaller batch size? I'm not very familiar with RNN and CTC but if, I'm not mistaken, early papers on it trained the networks online (mini-batch of size 1) and it seemed to work. Even if the gradients are noisy, I think you'll be better off doing more noisy updates than many passes for a single, even if more precise, update.

If you want to proceed with this: [this is where Keras actually makes the backend function that will compute the loss and update parameters](https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L700). I'm not entirely sure how you would efficiently hack this in, but you could start from there.
